Hide AVX512 by default on SKX/CLX/CPX

It can be enabled by providing "cpuid=avx512" in Xen's cmdline.

Signed-off-by: Sergey Dyasli <sergey.dyasli@citrix.com>
Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>

diff --git a/xen/arch/x86/cpu-policy.c b/xen/arch/x86/cpu-policy.c
index 8ecbc3adc884..4fa61264477b 100644
--- a/xen/arch/x86/cpu-policy.c
+++ b/xen/arch/x86/cpu-policy.c
@@ -45,6 +45,8 @@ static const struct feature_name {
     unsigned int bit;
 } feature_names[] __initconstrel = INIT_FEATURE_NAMES;
 
+int8_t __initdata opt_avx512 = -1;
+
 /*
  * Parse a list of cpuid feature names -> bool, calling the callback for any
  * matches found.
@@ -66,6 +68,12 @@ static int __init always_inline parse_cpuid(
         if ( !ss )
             ss = strchr(s, '\0');
 
+        if ( (val = parse_boolean("avx512", s, ss)) >= 0 )
+        {
+            opt_avx512 = val;
+            goto next;
+        }
+
         /* Skip the 'no-' prefix for name comparisons. */
         feat = s;
         if ( strncmp(s, "no-", 3) == 0 )
@@ -109,6 +117,7 @@ static int __init always_inline parse_cpuid(
         if ( mid )
             rc = -EINVAL;
 
+    next:
         s = ss + 1;
     } while ( *ss );
 
diff --git a/xen/arch/x86/include/asm/cpuid.h b/xen/arch/x86/include/asm/cpuid.h
index b32ba0bbfe5c..412376cf1833 100644
--- a/xen/arch/x86/include/asm/cpuid.h
+++ b/xen/arch/x86/include/asm/cpuid.h
@@ -10,6 +10,8 @@
 
 #include <public/sysctl.h>
 
+extern int8_t opt_avx512;
+
 extern const uint32_t known_features[FSCAPINTS];
 
 /*
diff --git a/xen/arch/x86/setup.c b/xen/arch/x86/setup.c
index 4be3fb846b10..8a27ed11ba5b 100644
--- a/xen/arch/x86/setup.c
+++ b/xen/arch/x86/setup.c
@@ -1853,6 +1853,18 @@ void __init noreturn __start_xen(unsigned long mbi_p)
     if ( cpu_has_pks )
         wrpkrs_and_cache(0); /* Must be before setting CR4.PKS */
 
+    /*
+     * The AVX512 implementation on SKX/CLX/CPX has package-wide frequency
+     * implications, so is turned off by default.  ICX is far better, and
+     * other vendors have no frequency implications at all.
+     */
+    if ( opt_avx512 == -1 )
+        opt_avx512 = !(boot_cpu_data.x86_vendor == X86_VENDOR_INTEL &&
+                       boot_cpu_data.x86 == 0x6 &&
+                       boot_cpu_data.x86_model == 0x55);
+    if ( !opt_avx512 )
+        setup_clear_cpu_cap(X86_FEATURE_AVX512F);
+
     init_speculation_mitigations();
 
     init_idle_domain();
